### DISCUSSION TITLE
Copy different files when using SIMULATION=native

### DISCUSSION THREAD

**Discussion by the-other-james** [nasa/cFS#419] [Q&A]
Our cFS project has configuration files that we copy using the FILELIST in [targets.cmake](https://github.com/nasa/cFE/blob/main/cmake/sample_defs/targets.cmake#L38). This works fine when we have one target as the native build and another target for the actual hardware (i.e. cpu1 is our native build, cpu2 is for the target hardware). But this breaks when we try to use `SIMULATION=native` for native builds. I was hoping that the build system would pickup on files with "native" as a prefix (similar to `native_osconfig.cmake`), but this isn't the case.

For example, in targets.cmake we set:
`SET(cpu1_FILELIST our_file.txt)`

In our mission defs, we have "cpu1_our_file.txt" and "native_our_file.txt". If we run `make prep`, the correct file is selected (cpu1_our_file.txt). If we run `make SIMULATION=native prep` then the wrong file is selected (expected native_our_file.txt, selected cpu1_our_file.txt).

This is somewhat related to https://github.com/nasa/cFS/discussions/403... and I think we could hack something together that would work, but I wanted to see if there is an easy fix or a better way to handle this.

Thanks!

**Comment by skliper** [comment]
TBH I think it would be really good if we could implement a standard cmake function to do matching/overrides from within *_defs to bring in the desired element.  There's a few different matching patterns used now based on what it's matching... sounds like the file pattern match has something unique.

**Reply by the-other-james** [reply]
That sounds like a good change to me. Right now the file matching is done here: https://github.com/nasa/cFE/blob/main/cmake/arch_build.cmake#L703

For now I'm going to add a clause such as:

```
      if (SIMULATION)
        message("We are living in a simulation: native_${INSTFILE}")
        if(EXISTS ${MISSION_DEFS}/native_${INSTFILE})
            set(FILESRC ${MISSION_DEFS}/native_${INSTFILE})
        endif()
```

That doesn't cover all of the desired cases... I'm thinking that for builds with multiple cpus you would want to be able to have native_cpu1_filename.ext, native_cpu2_filename_ext, etc.

**Comment by jphickey** [comment]
The "FILELIST" is a little too simplistic for a lot of use cases.

Recommendation is to use the `install_custom.cmake` hook, which is included/sourced at the same time the FILELIST is processed. In this you can use any CMake conditional/variable logic and call `install` directly (with all options).  Basically, you can write CMake code to do whatever you need, install as many files as you want, with whatever name translation you want....
